lib/integration/esbuild.js

```diff
@@ -2,127 +2,206 @@
-    /** @type {Point} */
-    let end
-    /** @type {number} */
-    let length
-    /** @type {number} */
-    let lineStart
-    /** @type {number} */
-    let lineEnd
-    /** @type {RegExpExecArray} */
-    let match
-    /** @type {number} */
-    let line
-    /** @type {number} */
-    let column
-
-    try {
-      file = await process(file)
-      contents = file.contents
-      messages = file.messages
-    } catch (error) {
-      error.fatal = true
-      messages.push(error)
+    if (allowDangerousRemoteMdx) {
+      // Intercept import paths starting with "http:" and "https:" so
+      // esbuild doesn't attempt to map them to a file system location.
+      // Tag them with the "http-url" namespace to associate them with
+      // this plugin.
+      build.onResolve(
+        {filter: filterHttp, namespace: 'file'},
+        resolveRemoteInLocal
+      )
+
+      build.onResolve(
+        {filter: filterHttpOrRelative, namespace: remoteNs},
+        resolveInRemote
+      )
     }
 
-    for (message of messages) {
-      start = message.location.start
-      end = message.location.end
-      length = 0
-      lineStart = 0
-      line = undefined
-      column = undefined
-
-      if (start.line != null && start.column != null && start.offset != null) {
-        line = start.line
-        column = start.column - 1
-        lineStart = start.offset - column
-        length = 1
-
-        if (end.line != null && end.column != null && end.offset != null) {
-          length = end.offset - start.offset
-        }
+    build.onLoad({filter: /.*/, namespace: remoteNs}, onloadremote)
+    build.onLoad({filter}, onload)
+
+    /** @param {OnResolveArgs} args  */
+    function resolveRemoteInLocal(args) {
+      return {path: args.path, namespace: remoteNs}
+    }
+
+    // Intercept all import paths inside downloaded files and resolve them against
+    // the original URL. All of these
+    // files will be in the "http-url" namespace. Make sure to keep
+    // the newly resolved URL in the "http-url" namespace so imports
+    // inside it will also be resolved as URLs recursively.
+    /** @param {OnResolveArgs} args  */
+    function resolveInRemote(args) {
+      return {
+        path: String(new URL(args.path, args.importer)),
+        namespace: remoteNs
       }
+    }
 
-      eol.lastIndex = lineStart
-      match = eol.exec(doc)
-      lineEnd = match ? match.index : doc.length
-      ;(message.fatal ? errors : warnings).push({
-        pluginName: name,
-        text: message.reason,
-        notes: [],
-        location: {
-          namespace: 'file',
-          suggestion: '',
-          file: data.path,
-          line,
-          column,
-          length: Math.min(length, lineEnd),
-          lineText: doc.slice(lineStart, lineEnd)
-        },
-        detail: message
-      })
+    /**
+     * @param {OnLoadArgs} data
+     * @returns {Promise<OnLoadResult>}
+     */
+    async function onloadremote(data) {
```

Comment:
```suggestion
    async function onloadremote({path}) {
```

Reply:
Not a fan of destructuring *in* params,
- The error is hard to read if a weird value is given: `Cannot destructure property 'path' of 'object null' as it is null`
- hard to figure out / log what else was given
- I’m fine with defaults *in* params, or destructuring after (e.g., [here](https://github.com/wooorm/xdm/pull/66/files#diff-59d6233a7bb3ebea2d25c8defe17f770568246e1d793ee59b9caa4368dc647abR36)), but no need to store stuff in variables for these three imo